tags: initial implemention of post tagging

taskboy3000 opened the pull request:
This patch adds the facility to plerd for tagging blog posts and
creating tag indexes of same.

* new attribute 'tags' supported in blog post source files
* tags are a comma-seperated list (quoting is not required or
  supported)
* tag index added to nav bar in wrapper.tt
* new tag index view shows all tags and their associated posts
* tag NAME specific views are created for each tag
* within templates, plerd.tag_uri(TAGNAME) can be use to get URI for a
  given tag
* updated unit test to support tags

taskboy3000 committed:
tags: initial implemention of post tagging

This patch adds the facility to plerd for tagging blog posts and
creating tag indexes of same.

* new attribute 'tags' supported in blog post source files
* tags are a comma-seperated list (quoting is not required or
  supported)
* tag index added to nav bar in wrapper.tt
* new tag index view shows all tags and their associated posts
* tag NAME specific views are created for each tag
* within templates, plerd.tag_uri(TAGNAME) can be use to get URI for a
  given tag
* updated unit test to support tags

taskboy3000 committed:
tags with spaces: spaces replaced with underscores

This change ensures that tags with spaces are written to disk with
underscores.  plerd->tag_uri() does the right thing.

jmacdotorg commented on lib/Plerd/Post.pm:
```diff
@@ -512,7 +525,9 @@ sub _process_source_file {
     if ( $attributes_need_to_be_written_out ) {
         my $new_content = '';
         for my $attribute_name ( @ordered_attribute_names ) {
-            $new_content .= "$attribute_name: $attributes{ $attribute_name }\n";
```
This change seems unrelated to the tag-specific work?

taskboy3000 replied:
This change was needed to silence warnings in the test.  This wasn't a delete, but an existence/defined check (see below).   There is no harm input validation, is there?

jmacdotorg replied:
Okay, makes sense, thanks.

May as well just collapse the if-check to `defined $attributes{ $attribute_name }` though, right?

jmacdotorg commented on t/basic.t:
```diff
@@ -54,8 +54,9 @@ unlink "$FindBin::Bin/source/no-title.md";
 
-# The "+3" below accounts for the generated recent, archive, and RSS files.
-my $expected_docroot_count = scalar( $source_dir->children( no_hidden => 1 ) ) + 3;
+# The "+3" below accounts for the generated recent, archive, tag
```
It's +4, now. :)

jmacdotorg replied:
(Also, the list of stuff makes it sound like it should be +6?)

jmacdotorg commented on t/templates/_tag.tt:
```diff
@@ -0,0 +1,8 @@
+<section>
```
I would rather that this file just be a [% BLOCK %] section in tags.tt, the only template that invokes it.

Using a sub-template file like this isn't bad at all, but making it into a TT-block would be more consistent with the existing example templates.

jmacdotorg commented on templates/post.tt:
```diff
@@ -125,8 +135,3 @@
 [% END %]
-
-
-<style>
```
Why remove this?

jmacdotorg commented on t/templates/wrapper.tt:
```diff
@@ -3,11 +3,23 @@
     <meta name="viewport" content="width=device-width, initial-scale=1.0">
-    <link href="http://netdna.bootstrapcdn.com/bootstrap/3.0.3/css/bootstrap.min.css" rel="stylesheet">
-    <link href="atom.xml" rel="alternate" title="[% plerd.title %]" type="application/atom+xml">
+    <link href="https://netdna.bootstrapcdn.com/bootstrap/3.0.3/css/bootstrap.min.css" rel="stylesheet">
```
Why add all this to the t/ templates now?

(It may be a good idea, but it's unrelated to the tag work...)

jmacdotorg replied:
Oh, it looks like you pasted in wrapper.tt from templates/wrapper.tt.

I don't think that's at all necessary just to make the tests work, right? Go ahead and just revert this file, please.

jmacdotorg commented on t/templates/post.tt:
```diff
@@ -110,7 +110,3 @@
     [% END %]
 [% END %]
 
-
```
Removing this is probably a good idea, but it's not related to the tag work.

jmacdotorg commented on templates/tags.tt:
```diff
@@ -0,0 +1,20 @@
+
+[%   END %]
+
+[% END %]
```
See comments on _tag.tt about adding  [% BLOCK %] to this file.

jmacdotorg commented on templates/_tag.tt:
```diff
@@ -0,0 +1,8 @@
+<section>
+    <h3><a href="[% plerd.tag_uri(tag) %]">[% tag %]</a></h3>
```
As written, this causes a tag-specific page to link to itself via its sole header, which is uneccesary.

jmacdotorg commented on templates/tags.tt:
```diff
@@ -0,0 +1,20 @@
+[% WRAPPER wrapper.tt title = 'Tags' %]
```
The tag-specific pages look good, but I'm not sure about the top-level tag index. Will post thoughts on the overall review.

jmacdotorg commented on lib/Plerd.pm:
```diff
@@ -569,6 +650,70 @@ sub generates_post_guids {
+
+# Return either the tags/index.html file
+# or a tags/TAGNAME.html file if given a tag
+sub tags_publication_file {
```
All these public methods ought to be documented in the appropriate section of the POD at the end of this file.

jmacdotorg commented on lib/Plerd.pm:
```diff
@@ -40,13 +40,26 @@ has 'database_path' => (
     isa => 'Str',
 );
 
+has 'tags_publication_path' => (
```
All these public attributes ought to be documented in the appropriate section of the POD at the end of this file.

jmacdotorg commented on lib/Plerd/Post.pm:
```diff
@@ -60,6 +60,12 @@ has 'attributes' => (
     isa => 'HashRef',
 );
 
+has 'tags' => (
```
This attribute ought to be documented in the appropriate section of the POD at the end of this file.

jmacdotorg reviewed (changes requested):
This is great! I'm requesting a pile of changes but they're mostly little stuff.

One heavier bit: I think I'd prefer the top-level tag index page to be list a simple list of tags, maybe with a parenthesized count of the matching posts next to each, rather than a fully expanded list of every tag's posts. That would open the door to many (even most) posts getting listed again and again on that page, and I don't like that thought. A nice clean list seems better.

(The tag-specific pages seem fine though!)

jmacdotorg commented:
Two more comments:

1. The tags' HTML should use `p-category` Microformats2 markup, per <http://microformats.org/wiki/h-entry>.

1. Links to pages with punctuation or emoji don't work, though the tag-index pages for them seem to be created okay, with URI-encoded names. For example: `😀` or `"Foo bar"`. (There should be at least one test for these, as well.)

taskboy3000 commented:
1. I thought the blog posts were markup correctly using the h-entry stuff.  Didn't you point me to a validator?  However, I think I should just revert the wrapper.  I admit that I blanket copied the templates.

jmacdotorg commented:
Yes, the blog posts validate as `h-entry`s. I suggest that we take the opportunity of adding tags to further enrich the existing Microformats metadata, since `h-entry` has a property called `p-category` tailor-made for this.

So, we could e.g. change post.tt so that tags are formatted this way (adding only one new class to the code you've already written):

`<li class="list-inline-item p-category"><a href="[% plerd.tag_uri(tag) %]">[% tag %]</a></li>`

Make sense?

taskboy3000 commented:
Yes, that makes sense.  Will tag the, er, tags.

--Joe

> On Aug 21, 2018, at 9:33 PM, Jason McIntosh <notifications@github.com> wrote:
> 
> Yes, the blog posts validate as h-entrys. I suggest that we take the opportunity of adding tags to further enrich the existing Microformats metadata, since h-entry has a property called p-category tailor-made for this.
> 
> So, we could e.g. change post.tt so that tags are formatted this way (adding only one new class to the code you've already written):
> {{{
> 
> [% tag %]
> }}} Make sense?
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

taskboy3000 commented:
I thought that in some old Perl, that defined check would autovivicate the key.  I can’t reproduce that now, so it seems harmless to reduce the check to simply defined.

--Joe

> On Aug 21, 2018, at 10:11 PM, Jason McIntosh <notifications@github.com> wrote:
> 
> @jmacdotorg commented on this pull request.
> 
> In lib/Plerd/Post.pm:
> 
> > @@ -512,7 +525,9 @@ sub _process_source_file {
>      if ( $attributes_need_to_be_written_out ) {
>          my $new_content = '';
>          for my $attribute_name ( @ordered_attribute_names ) {
> -            $new_content .= "$attribute_name: $attributes{ $attribute_name }\n";
> Okay, makes sense, thanks.
> 
> May as well just collapse the if-check to defined $attributes{ $attribute_name } though, right?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

taskboy3000 committed:
tags: removed _tag.tt; enhanced tags.tt

This patch reworks the tag index views.

Issues addressed: 2, 8, 5

* _tag.tt fragment removed
* tags index view
** removed all references to _tag.tt fragment
** tag index view no longer lists posts, but displays # of posts for
   that tag
** tag-specific view no longer links the header to itself
* these same changes were applied to the templated within t/

taskboy3000 committed:
tags: revert t/templates/post.tt, t/templates/wrapper.tt

This patch reverts the changes to the templates in t/.

The contents were pulled from plerd master

Issues addressed: 6, 7

taskboy3000 committed:
tag: fixed comment in t/basic.t

This patch fixes a typo in comment.

Issue addressed: 4

taskboy3000 committed:
tags: added 'p-category' to tags in post.tt

This patch adds the 'p-category' to tags in the post view.  This
better supports Microformats2 markup.

Issue addressed: 11

taskboy3000 committed:
tags: updated POD about public methods

This patch enhances the exisiting POS of these classes with
information about public classes that affect tags.

Issues addressed: 9, 10

taskboy3000 committed:
tags: removed unneeded existence check from Post.pm

This patch removes an unneed existence check from Plerd::Post.
Concerns about autovivication in the case of a simple hash were
spurious.

taskboy3000 committed:
tags: fixing tags template

This patch makes the tags index page display tag links and post counts
properly.

taskboy3000 committed:
tags: specific tag index view fix

This patch fixes the problem where the view did not iterate over the
posts associated with the tag.

jmacdotorg commented on lib/Plerd.pm:
```diff
@@ -913,6 +918,21 @@ the blog's docroot -- in other words, the place Plerd will write HTML and XML fi
+This is a L<URI> object that points to the tag index.  It is
+particularly helpful when creating navigation.
+
+=item tag_uri TAG
```
Please move this to the "OBJECT METHODS" list instead, and format the argument list per the other argument-taking methods found there.

jmacdotorg commented:
Holding off further work on this until #28 (and related tasks) are resolved.

jmacdotorg commented:
On @taskboy3000's request, I'm taking over further development of this branch. (I plan to request his approval on significant changes I make prior to any merge into the master branch.)

jmacdotorg commented:
Moving this work into local branch https://github.com/jmacdotorg/plerd/tree/tags. Will open a new pull request based off that after some further field-testing.
